packages/providers/src/transaction-response/transaction-response.ts

```diff
@@ -0,0 +1,132 @@
+/* eslint-disable @typescript-eslint/no-non-null-assertion */
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import { arrayify } from '@ethersproject/bytes';
+import type {
+  ReceiptCall,
+  ReceiptLog,
+  ReceiptLogData,
+  ReceiptPanic,
+  ReceiptReturn,
+  ReceiptReturnData,
+  ReceiptRevert,
+  ReceiptTransfer,
+  ReceiptTransferOut,
+  ReceiptScriptResult,
+} from '@fuel-ts/transactions';
+import { ReceiptType, ReceiptCoder } from '@fuel-ts/transactions';
+
+import type {
+  GqlGetTransactionWithReceiptsQuery,
+  GqlReceiptFragmentFragment,
+} from '../__generated__/operations';
+import type Provider from '../provider';
+import type { TransactionRequest } from '../transaction-request';
+
+export type TransactionResultReceipt =
+  | ReceiptCall
+  | ReceiptReturn
+  | (ReceiptReturnData & { data: string })
+  | ReceiptPanic
+  | ReceiptRevert
+  | ReceiptLog
+  | (ReceiptLogData & { data: string })
+  | ReceiptTransfer
+  | ReceiptTransferOut
+  | ReceiptScriptResult;
+
+export type TransactionResult<TStatus extends 'success' | 'failure'> = {
+  status: TStatus extends 'success'
+    ? { type: 'success'; programState: any }
+    : { type: 'failure'; reason: any };
+  /** Receipts produced during the execution of the transaction */
+  receipts: TransactionResultReceipt[];
+  blockId: any;
+  time: any;
+};
+
+const processGqlReceipt = (gqlReceipt: GqlReceiptFragmentFragment): TransactionResultReceipt => {
+  const receipt = new ReceiptCoder('receipt').decode(arrayify(gqlReceipt.rawPayload), 0)[0];
+
+  switch (receipt.type) {
+    case ReceiptType.ReturnData: {
+      return {
+        ...receipt,
+        data: gqlReceipt.data!,
+      };
+    }
+    case ReceiptType.LogData: {
+      return {
+        ...receipt,
+        data: gqlReceipt.data!,
+      };
+    }
+    default:
+      return receipt;
+  }
+};
+
+export class TransactionResponse {
+  /** Transaction ID */
+  id: string;
+  /** Transaction request */
+  request: TransactionRequest;
+  provider: Provider;
+
+  constructor(id: string, request: TransactionRequest, provider: Provider) {
+    this.id = id;
+    this.request = request;
+    this.provider = provider;
+  }
+
+  async #fetch(): Promise<NonNullable<GqlGetTransactionWithReceiptsQuery['transaction']>> {
+    const { transaction } = await this.provider.operations.getTransactionWithReceipts({
+      transactionId: this.id,
+    });
+    if (!transaction) {
+      throw new Error('No Transaction was received from the client.');
+    }
+    return transaction;
+  }
+
+  /** Waits for transaction to succeed or fail and returns the result */
+  async waitForResult(): Promise<TransactionResult<any>> {
+    const transaction = await this.#fetch();
+
+    switch (transaction.status?.type) {
+      case 'SubmittedStatus': {
+        // TODO: Implement polling or GQL subscription
```

Comment:
Can you create an issue for this?

Reply:
We don't need to track this separately since it will be a mandatory change when we will eventually need to upgrade to a fuel-core version that doesn't immediately produce blocks.